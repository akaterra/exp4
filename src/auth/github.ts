import express from 'express';
import { Service } from 'typedi';
import { EntityService } from '../entities.service';
import { Autowired, err, request } from '../utils';
import { IAuthStrategyMethod, IAuthStrategyService } from './auth-strategy.service';
import { IUser } from '../user';
import { StoragesService } from '../storages.service';
import { authSendData as execAuthSendData, prepareAuthData } from '../auth.service';
import { Log } from '../logger';

@Service()
export class GithubAuthStrategyService extends EntityService implements IAuthStrategyService {
  static readonly type: string = 'github';

  @Autowired() protected storagesService: StoragesService;

  private get storage() {
    return this.storagesService.get(this.config?.storage ?? 'default');
  }

  constructor(protected config?: {
    credentials?: {
      clientId?: string;
      clientSecret?: string;
      callbackUrl?: string;
    };
    storage?: string;
  }) {
    super();

    if (config) {
      if (!this.config.credentials) {
        this.config.credentials = {};
      }

      if (!this.config.credentials.clientId) {
        this.config.credentials.clientId = process.env.GITHUB_CLIENT_ID;
      }

      if (!this.config.credentials.clientSecret) {
        this.config.credentials.clientSecret = process.env.GITHUB_CLIENT_SECRET;
      }

      if (!this.config.credentials.callbackUrl) {
        this.config.credentials.callbackUrl = process.env.GITHUB_CALLBACK_URL;
      }
    }
  }

  @Log('debug')
  async authorize(data: Record<string, any>): Promise<IUser> { // eslint-disable-line
    return null;
  }

  @Log('debug')
  async request(): Promise<IAuthStrategyMethod> {
    return {
      id: this.id,
      type: this.type,
      actions: {
        redirect: `https://github.com/login/oauth/authorize?client_id=${this.config?.credentials?.clientId}&client_secret=${this.config?.credentials?.clientSecret}&redirect_uri=${encodeURIComponent(this.getRedirectUri())}`,
      },
    };
  }

  @Log('debug')
  async configureServer(app: express.Application, path?: string): Promise<void> {
    if (!path) {
      path = `/auth/methods/${this.id ?? this.type}`;
    } else {
      path = `/auth/methods${path}`;
    }

    app.get(path, err(async (req, res) => {
      res.json(await this.request());
    }));

    app.get(path + '/redirect', err(async (req, res) => {
      res.redirect((await this.request()).actions.redirect);
    }));

    app.get(path + '/callback', err(async (req, res) => {
      const githubAuth = await request(
        'https://github.com/login/oauth/access_token',
        {
          client_id: this.config?.credentials?.clientId,
          client_secret: this.config?.credentials?.clientSecret,
          code: req.query.code,
          redirect_uri: this.getRedirectUri(),
        },
        'post',
      );
      const githubUser = await request(
        'https://api.github.com/user',
        undefined,
        'get',
        githubAuth.access_token,
      );

      const user = this.config?.storage
        ? await this.storage.userGetById(String(githubUser.id), this.type)
          ?? (githubUser.email && await this.storage.userGet({ email: githubUser.email, type: this.type }))
          ?? (githubUser.login && await this.storage.userGet({ login: githubUser.login, type: this.type }))
          ?? null
        : {
          id: String(githubUser.id),
          type: this.type,
  
          // name: authData.name,
          email: githubUser.email,
        };

      if (this.config?.storage && !user) {
        res.status(401);

        throw new Error('User not found');
      }

      execAuthSendData(req, res, prepareAuthData(user));
    }));
  }

  private getRedirectUri() {
    return this.config?.credentials?.callbackUrl || `${process.env.DOMAIN || 'http://localhost:9002'}/auth/${this.id}/callback`;
  }
}
