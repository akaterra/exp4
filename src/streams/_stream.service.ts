import { IStreamStateContext, StreamState } from '../stream';
import { IProjectTargetDef, IProjectTargetStreamDef } from '../project';
import { TargetState } from '../target';
import { IService } from '../entities.service';

export enum StreamServiceStreamMoveOptsStrategy {
  APPROVE = 'approve',
  REQUEST = 'request',
}

export interface IStreamServiceStreamMoveOpts {
  strategy?: StreamServiceStreamMoveOptsStrategy;
}

export interface IStreamService extends IService {
  actionRun(id: string);

  streamBookmark(stream: IProjectTargetStreamDef): Promise<StreamState>;

  streamDetach(stream: IProjectTargetStreamDef): Promise<StreamState>;

  streamGetState(stream: IProjectTargetStreamDef, scopes?: Record<string, boolean>, context?: IStreamStateContext): Promise<StreamState>;

  streamMove(sourceStream: IProjectTargetStreamDef, targetStream: IProjectTargetStreamDef, opts?: IStreamServiceStreamMoveOpts);

  targetGetState(target: IProjectTargetDef, scopes?: Record<string, boolean>): Promise<TargetState>;
}
