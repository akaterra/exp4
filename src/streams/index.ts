import { IStreamStateContext, StreamState } from '../stream-state';
import { IProjectTargetDef, IProjectTargetStreamDef } from '../project';
import { TargetState } from '../target-state';
import { IService } from '../entities.service';
import { Service } from 'typedi';
import { AwaitedCache, Mutex } from '../cache';
import { ProjectsService } from '../projects.service';
import { EntitiesServiceWithFactory } from '../entities.service';
import { Autowired, CallbacksContainer } from '../utils';
import { logError } from '../logger';
import {EVENT_STREAM_STATE_REREAD_FINISHED, EVENT_STREAM_STATE_REREAD_STARTED, EVENT_TARGET_STATE_REREAD_STARTED} from '../const';

export enum StreamServiceStreamMoveOptsStrategy {
  APPROVE = 'approve',
  REQUEST = 'request',
}

export interface IStreamServiceStreamMoveOpts {
  strategy?: StreamServiceStreamMoveOptsStrategy;
}

export interface IStreamService extends IService {
  actionRun(id: string);

  streamBookmark(stream: IProjectTargetStreamDef): Promise<StreamState>;

  streamDetach(stream: IProjectTargetStreamDef): Promise<StreamState>;

  streamGetState(stream: IProjectTargetStreamDef, scopes?: Record<string, boolean>, context?: IStreamStateContext): Promise<StreamState>;

  streamMove(sourceStream: IProjectTargetStreamDef, targetStream: IProjectTargetStreamDef, opts?: IStreamServiceStreamMoveOpts);

  targetGetState(target: IProjectTargetDef, scopes?: Record<string, boolean>): Promise<TargetState>;
}

@Service()
export class StreamHolderService extends EntitiesServiceWithFactory<IStreamService> {
  @Autowired(() => ProjectsService) protected projectsService: ProjectsService;
  protected cache = new AwaitedCache<StreamState>();
  protected mutex = new Mutex();

  get domain() {
    return 'Stream';
  }

  constructor(public readonly callbacksContainer: CallbacksContainer = new CallbacksContainer()) {
    super();
  }

  async rereadState(stream: IProjectTargetStreamDef, scopes?: Record<string, boolean>, context?: IStreamStateContext): Promise<StreamState> {
    context = context ? { ...context } : {};

    const project = this.projectsService.get(stream.ref?.projectId);
    const key = `${stream.ref.projectId}:${stream.ref.targetId}:${stream.id}`;
    const release = await this.mutex.acquire(key);

    try {
      const streamState = stream.isDirty || scopes
        ? await this.get(stream.type).streamGetState(stream, scopes, context)
        : await this.cache.get(key) ?? await this.get(stream.type).streamGetState(stream, scopes, context);

      if (!streamState) {
        return null;
      }

      await this.callbacksContainer.run(
        EVENT_STREAM_STATE_REREAD_STARTED,
        { stream, streamState },
      );

      streamState.stream = stream;

      if (context.artifact) {
        try {
          streamState.isSyncing = true;

          await project.env.artifacts.run(
            { artifacts: stream.artifacts, ref: stream.ref },
            streamState,
            context.artifact,
            scopes,
          );
        } catch (err) {
          logError(err, 'StreamsService.rereadState', { ref: stream.ref, scopes });
        } finally {
          streamState.isSyncing = false;
        }
      }

      streamState.version = streamState.version ?? await project
        .getEnvVersioningByTargetStream(stream)
        .getCurrent(project.getTargetByTargetStream(stream));

      stream.isDirty = false;

      this.cache.set(key, streamState);
      streamState.isDirty = false;

      await this.callbacksContainer.run(
        EVENT_STREAM_STATE_REREAD_FINISHED,
        { stream, streamState },
      );

      return streamState;
    } finally {
      release();
    }
  }
}
