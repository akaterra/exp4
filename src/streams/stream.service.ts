import { IStream } from '../stream';
import { IProjectTarget, IProjectTargetStreamDef } from '../project';
import { ITarget } from '../target';
import { IService } from '../entities.service';

export interface IStreamService extends IService {
  actionRun(id: string);

  streamDetach(stream: IProjectTargetStreamDef): Promise<IStream>;

  streamGetState(stream: IProjectTargetStreamDef): Promise<IStream>;

  streamMove(sourceStream: IProjectTargetStreamDef, targetStream: IProjectTargetStreamDef);

  targetGetState(target: IProjectTarget): Promise<ITarget>;
}
