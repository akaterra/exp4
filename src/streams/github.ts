import { Octokit } from '@octokit/rest';
import { IStreamService } from './stream.service';
import { IProjectTarget, IProjectTargetStream } from '../project';
import { IStream } from '../stream';
import { Service } from 'typedi';
import { ITarget } from '../target';
import { EntityService } from '../entities.service';
import { Autowired, hasScope } from '../utils';
import { GithubIntegrationService } from '../integrations/github';
import { ProjectsService } from '../projects.service';
import { Status } from '../enums/status';
import { AwaitedCache } from '../cache';

const JOB_CONSLUSION_TO_STATUS_MAP = {
  failure: Status.FAILED,
  skipped: Status.COMPLETED,
  success: Status.COMPLETED,
}
const JOB_STATUS_TO_STATUS_MAP = {
  is_progress: Status.PROCESSING,
}

export type IGithubTargetStream = IProjectTargetStream<{
  integration?: string;
  org: string;
  branch: string;
}, 'github'>;

export type IGithubStream = IStream<{
  sha: string;
  branch: string;
}>;

@Service()
export class GithubStreamService extends EntityService implements IStreamService {
  static readonly type: string = 'github';

  protected cache = new AwaitedCache<IStream>();
  protected client = new Octokit({
    auth: process.env.GITHUB_TOKEN,
  });

  actionRun(id: string) { // eslint-disable-line

  }

  async streamDetach(stream: IGithubTargetStream): Promise<IStream> {
    const integration = this.getIntegrationService(stream);
    const branchName = await this.getBranch(stream);

    await integration.branchDelete(
      branchName,
      stream.id,
    );

    return null;
  }

  async streamGetState(stream: IGithubTargetStream, scopes?: Record<string, boolean>): Promise<IStream> {
    const cacheKey = `${stream.ref?.projectId}:${stream.ref?.targetId}:${stream.ref?.streamId}`;
    const state: IStream = await this.cache.get(cacheKey) ?? {
      id: stream.id,
      type: this.type,

      projectId: stream.ref.projectId,
      targetId: stream.ref.targetId,

      history: {
        action: [],
        artifact: [],
        change: [],
      },
      isSyncing: true,
      link: null,
      metadata: {},
      version: null,
    };

    const detailsPromise = (async () => {
      state.isSyncing = true;

      const integration = this.getIntegrationService(stream);
      const branchName = await this.getBranch(stream);
      const branch = hasScope('change', scopes) ? await integration.gitGetBranch(branchName, stream.id) : null;
      const versioningService = await this.projectsService
        .get(stream.ref.projectId)
        .getEnvVersioningByTargetId(stream.ref.targetId);
      const workflowRuns = hasScope('action', scopes) && branch
        ? await integration.gitGetWorkflowRuns(stream.config.branch, stream.id)
        : null;
      const workflowRunsJobs = hasScope('action', scopes) && workflowRuns?.[0] && workflowRuns?.[0]?.id !== parseInt(state.history?.action?.[0]?.id)
        ? await integration.gitGetWorkflowJobs(workflowRuns[0].id, stream.id, stream.config.org)
        : null;
  
      const metadata = {
        org: stream.config?.org ?? integration.config?.org,
        branch: branchName,
      };

      if (hasScope('action', scopes)) {
        state.history.action = workflowRuns?.length ? workflowRuns.map((w) => {
          const isJobStepsNotFailed = workflowRunsJobs?.[0]
            ? workflowRunsJobs[0].steps.every((jobStep) => jobStep.conclusion !== 'failure')
            : null;

          return {
            id: String(w.id),
            type: 'github:workflow',
    
            author: {
              name: w.actor?.name ?? w.actor?.login ?? null,
              link: w.actor?.html_url ?? null,
            },
            description: w.name,
            link: w.html_url ?? null,
            metadata: {},
            steps: workflowRunsJobs?.[0]
              ? workflowRunsJobs[0].steps.reduce((acc, jobStep) => {
                acc[jobStep.number] = {
                  id: String(jobStep.number),
                  type: 'github:workflow:job',

                  description: jobStep.name,
                  link: workflowRunsJobs[0].html_url,
                  status: JOB_CONSLUSION_TO_STATUS_MAP[jobStep.conclusion] ?? Status.UNKNOWN,
                };

                return acc;
              }, {})
              : state.history.action?.[0]?.steps ?? null,
            status: typeof isJobStepsNotFailed === 'boolean'
              ? isJobStepsNotFailed
                ? (JOB_STATUS_TO_STATUS_MAP[w.status] ?? w.status ?? null)
                : Status.FAILED
              : state.history.action?.[0]?.status ?? Status.COMPLETED,
            time: w.run_started_at,
          };
        }) : [];
      }

      if (hasScope('change', scopes)) {
        state.history.change = branch ? [ {
          id: branch.commit.sha,
          type: 'github:commit',

          author: {
            name: branch.commit.commit.author?.name ?? branch.commit.commit.author?.login ?? null,
            link: branch.commit.commit.author?.html_url ?? null,
          },
          description: branch.commit.commit.message,
          link: branch.commit.html_url,
          metadata: {},
          status: null,
          steps: null,
          time: branch.commit.commit.committer?.date ?? null,
        } ] : [];
      }

      state.link = branch ? branch._links.html : state?.link ?? null;
      state.metadata = metadata;
      state.version = await versioningService.getCurrentStream(stream);
  
      if (hasScope('artifact', scopes) && stream.artifacts?.length) {
        await this.getArtifactsService(stream).run(
          { artifacts: stream.artifacts, ref: stream.ref },
          state,
          {
            githubWorkflowRunJobId: workflowRunsJobs?.[0]?.id,
            githubWorkflowRunJobStatus: state.history.action?.[0]?.status,
            ref: stream.ref,
          },
        );
      }
    })().catch((err) => {
      console.error(err);
    }).finally(() => {
      state.isSyncing = false;
    });

    this.cache.set(cacheKey, state);

    if (stream.isDirty) {
      await detailsPromise;
    }

    return state;
  }

  async streamMove(sourceStream: IGithubTargetStream, targetStream: IGithubTargetStream) {
    const project = this.projectsService.get(sourceStream.ref?.projectId);

    project.env.streams.assertTypes(sourceStream.type, targetStream.type);

    const source = await this.projectsService
      .get(sourceStream.ref?.projectId)
      .getStreamStateByTargetIdAndStreamId(sourceStream.ref?.targetId, sourceStream.ref?.streamId, { change: true });

    if (!source?.history?.change?.[0]?.id) {
      return;
    }

    const sourceBranchName = await this.getBranch(sourceStream);
    const targetIntegration = this.projectsService
      .get(targetStream.ref.projectId)
      .getEnvIntegraionByTargetStream<GithubIntegrationService>(targetStream);
    const targetBranchName = await this.getBranch(targetStream);

    if (!await targetIntegration.gitGetBranch(
      targetBranchName,
      targetStream.id,
    )) {
      await targetIntegration.gitCreateReference(
        targetBranchName,
        source.history.change[0]?.id,
        targetStream.id,
      );
    }

    await targetIntegration.gitMerge(
      sourceBranchName,
      targetBranchName,
      `Release ${source.version}`,
      targetStream.id,
    )
  }

  async targetGetState(config: IProjectTarget): Promise<ITarget> {
    return {
      id: config.id,
      type: null,
    };
  }

  private getArtifactsService(stream: IGithubTargetStream) {
    return this.projectsService.get(stream.ref.projectId).env.artifacts;
  }

  private getIntegrationService(stream: IGithubTargetStream) {
    return this.projectsService
      .get(stream.ref.projectId)
      .getEnvIntegraionByTargetStream<GithubIntegrationService>(stream);
  }

  private async getBranch(stream: IGithubTargetStream) {
    const project = this.projectsService.get(stream.ref.projectId);
    const integration = project.getEnvIntegraionByTargetStream<GithubIntegrationService>(stream);
    const target = project.getTargetByTargetStream(stream);
    const branch = await project.getEnvVersioningByTarget(target)
      .getCurrent(target, stream.config?.branch ?? integration.config?.branch);

    return branch ?? integration?.config?.branch;
  }
}
