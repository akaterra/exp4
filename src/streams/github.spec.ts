import 'reflect-metadata';
import {IVersioningService} from '../versionings/versioning.service';
import {IProjectTargetDef, IProjectTargetStreamDef, Project} from '../project';
import {ProjectsService} from '../projects.service';
import {GithubStreamService} from './github';
import {StreamsService} from '../streams.service';
import Container from 'typedi';
import {IntegrationsService} from '../integrations.service';
import {GithubIntegrationService} from '../integrations/github';
import {StubVersioningService} from '../versionings/stub';
import {VersioningsService} from '../versionings.service';
import { Initial, Observer } from '@akaterra.co/unitsnap';

describe('Github stream', () => {
  const observer = new Observer();

  class TestGithubIntegrationService extends GithubIntegrationService {
    id = 'test';

    get type() {
      return 'test';
    }
  }

  class TestVersioningService extends StubVersioningService {
    id = 'test';

    get type() {
      return 'test';
    }
  }

  class TestProject extends Project {

  }

  function snapshot() {
    return observer
      .filter()
        .notPromiseResult()
      .snapshot()
        .addInstanceOfProcessor(TestGithubIntegrationService)
        .addInstanceOfProcessor(TestVersioningService)
        .includeName()
        .serialize();
  }

  beforeAll(() => {
    Container.get(ProjectsService).add(new TestProject({
      id: 'test',
      env: {
        integrations: new IntegrationsService().add(new TestGithubIntegrationService()),
        streams: new StreamsService(),
        versionings: new VersioningsService().add(new TestVersioningService()),
      },
    }));
  });

  beforeEach(() => {
    jest.restoreAllMocks();

    observer.begin();
  });

  afterEach(() => {
    observer.end();
  })

  it('should bookmark', async () => {
    const stream = new GithubStreamService();

    observer.override(TestGithubIntegrationService, {
      tagCreate: Function,
    });
    observer.override(TestProject, {
      getStreamStateByTargetIdAndStreamId: {
        history: {
          change: [ { id: 'changeId' } ],
        },
        version: 'version',
      },
    });

    await stream.streamBookmark({
      id: 'test',
      type: 'github',

      ref: {
        projectId: 'test',
      },

      config: {
        integration: 'test',
        org: 'org',
        repo: 'repo',
        branch: 'branch',
      },
    });

    expect(snapshot()).toMatchSnapshot();
  });

  it('should bookmark with stream id as repo', async () => {
    const stream = new GithubStreamService();

    observer.override(TestGithubIntegrationService, {
      tagCreate: Function,
    });
    observer.override(TestProject, {
      getStreamStateByTargetIdAndStreamId: {
        history: {
          change: [ { id: 'changeId' } ],
        },
        version: 'version',
      },
    });

    await stream.streamBookmark({
      id: 'test',
      type: 'github',

      ref: {
        projectId: 'test',
      },

      config: {
        integration: 'test',
        org: 'org',
        // repo: 'repo',
        branch: 'branch',
      },
    });

    expect(snapshot()).toMatchSnapshot();
  });

  it('should detach', async () => {
    const stream = new GithubStreamService();

    observer.override(TestGithubIntegrationService, {
      branchDelete: Function,
    });
    observer.override(TestProject, {
      getEnvIntegraionByTargetStream: Initial,
      getEnvVersioningByTarget: Initial,
      getTargetByTargetStream: {
        id: 'testTarget',
        type: 'test',

        versioning: 'test',
      },
    });
    observer.override(TestVersioningService, {
      getCurrent: Promise.resolve('testBranch'),
    });

    await stream.streamDetach({
      id: 'test',
      type: 'github',

      ref: {
        projectId: 'test',
      },

      config: {
        integration: 'test',
        org: 'org',
        repo: 'repo',
        branch: 'branch',
      },
    });

    expect(snapshot()).toMatchSnapshot();
  });

  it('should detach with stream id as repo', async () => {
    const stream = new GithubStreamService();

    observer.override(TestGithubIntegrationService, {
      branchDelete: Function,
    });
    observer.override(TestProject, {
      getEnvIntegraionByTargetStream: Initial,
      getEnvVersioningByTarget: Initial,
      getTargetByTargetStream: {
        id: 'testTarget',
        type: 'test',

        versioning: 'test',
      },
    });
    observer.override(TestVersioningService, {
      getCurrent: Promise.resolve('testBranch'),
    });

    await stream.streamDetach({
      id: 'test',
      type: 'github',

      ref: {
        projectId: 'test',
      },

      config: {
        integration: 'test',
        org: 'org',
        // repo: 'repo',
        branch: 'branch',
      },
    });

    expect(snapshot()).toMatchSnapshot();
  });
});
