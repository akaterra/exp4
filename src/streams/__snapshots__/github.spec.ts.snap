// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Github stream should bookmark 1`] = `
[
  {
    "args": {
      "*": [],
      "scopes": {
        "change": true,
      },
      "streamId": undefined,
      "targetId": undefined,
    },
    "name": "TestProject.getStreamStateByTargetIdAndStreamId",
  },
  {
    "name": "TestProject.getStreamStateByTargetIdAndStreamId",
    "result": {
      "history": {
        "change": [
          {
            "id": "changeId",
          },
        ],
      },
      "version": "version",
    },
  },
  {
    "args": {
      "*": [],
      "commitMessage": null,
      "lightweight": false,
      "org": null,
      "repo": "repo",
      "sha": "changeId",
      "tag": "version",
    },
    "name": "TestGithubIntegrationService.tagCreate",
  },
  {
    "name": "TestGithubIntegrationService.tagCreate",
  },
]
`;

exports[`Github stream should bookmark with stream id as repo 1`] = `
[
  {
    "args": {
      "*": [],
      "scopes": {
        "change": true,
      },
      "streamId": undefined,
      "targetId": undefined,
    },
    "name": "TestProject.getStreamStateByTargetIdAndStreamId",
  },
  {
    "name": "TestProject.getStreamStateByTargetIdAndStreamId",
    "result": {
      "history": {
        "change": [
          {
            "id": "changeId",
          },
        ],
      },
      "version": "version",
    },
  },
  {
    "args": {
      "*": [],
      "commitMessage": null,
      "lightweight": false,
      "org": null,
      "repo": "test",
      "sha": "changeId",
      "tag": "version",
    },
    "name": "TestGithubIntegrationService.tagCreate",
  },
  {
    "name": "TestGithubIntegrationService.tagCreate",
  },
]
`;

exports[`Github stream should detach 1`] = `
[
  {
    "args": {
      "*": [],
      "assertType": "github",
      "stream": {
        "config": {
          "branch": "branch",
          "integration": "test",
          "org": "org",
          "repo": "repo",
        },
        "id": "test",
        "ref": {
          "projectId": "test",
        },
        "type": "github",
      },
    },
    "name": "TestProject.getEnvIntegraionByTargetStream",
  },
  {
    "name": "TestProject.getEnvIntegraionByTargetStream",
    "result": {
      "$$data": "TestGithubIntegrationService",
      "$$type": "instanceOf",
    },
  },
  {
    "args": {
      "*": [],
      "assertType": "github",
      "stream": {
        "config": {
          "branch": "branch",
          "integration": "test",
          "org": "org",
          "repo": "repo",
        },
        "id": "test",
        "ref": {
          "projectId": "test",
        },
        "type": "github",
      },
    },
    "name": "TestProject.getEnvIntegraionByTargetStream",
  },
  {
    "name": "TestProject.getEnvIntegraionByTargetStream",
    "result": {
      "$$data": "TestGithubIntegrationService",
      "$$type": "instanceOf",
    },
  },
  {
    "args": {
      "*": [],
      "stream": {
        "config": {
          "branch": "branch",
          "integration": "test",
          "org": "org",
          "repo": "repo",
        },
        "id": "test",
        "ref": {
          "projectId": "test",
        },
        "type": "github",
      },
    },
    "name": "TestProject.getTargetByTargetStream",
  },
  {
    "name": "TestProject.getTargetByTargetStream",
    "result": {
      "id": "testTarget",
      "type": "test",
      "versioning": "test",
    },
  },
  {
    "args": {
      "*": [],
      "target": {
        "id": "testTarget",
        "type": "test",
        "versioning": "test",
      },
    },
    "name": "TestProject.getEnvVersioningByTarget",
  },
  {
    "name": "TestProject.getEnvVersioningByTarget",
    "result": {
      "$$data": "TestVersioningService",
      "$$type": "instanceOf",
    },
  },
  {
    "args": {
      "*": [
        "branch",
      ],
      "target": {
        "id": "testTarget",
        "type": "test",
        "versioning": "test",
      },
    },
    "name": "TestVersioningService.getCurrent",
  },
  {
    "name": "TestVersioningService.getCurrent",
    "result": "testBranch",
  },
  {
    "args": {
      "*": [],
      "branch": "testBranch",
      "repo": "repo",
    },
    "name": "TestGithubIntegrationService.branchDelete",
  },
  {
    "name": "TestGithubIntegrationService.branchDelete",
  },
]
`;

exports[`Github stream should detach with stream id as repo 1`] = `
[
  {
    "args": {
      "*": [],
      "assertType": "github",
      "stream": {
        "config": {
          "branch": "branch",
          "integration": "test",
          "org": "org",
        },
        "id": "test",
        "ref": {
          "projectId": "test",
        },
        "type": "github",
      },
    },
    "name": "TestProject.getEnvIntegraionByTargetStream",
  },
  {
    "name": "TestProject.getEnvIntegraionByTargetStream",
    "result": {
      "$$data": "TestGithubIntegrationService",
      "$$type": "instanceOf",
    },
  },
  {
    "args": {
      "*": [],
      "assertType": "github",
      "stream": {
        "config": {
          "branch": "branch",
          "integration": "test",
          "org": "org",
        },
        "id": "test",
        "ref": {
          "projectId": "test",
        },
        "type": "github",
      },
    },
    "name": "TestProject.getEnvIntegraionByTargetStream",
  },
  {
    "name": "TestProject.getEnvIntegraionByTargetStream",
    "result": {
      "$$data": "TestGithubIntegrationService",
      "$$type": "instanceOf",
    },
  },
  {
    "args": {
      "*": [],
      "stream": {
        "config": {
          "branch": "branch",
          "integration": "test",
          "org": "org",
        },
        "id": "test",
        "ref": {
          "projectId": "test",
        },
        "type": "github",
      },
    },
    "name": "TestProject.getTargetByTargetStream",
  },
  {
    "name": "TestProject.getTargetByTargetStream",
    "result": {
      "id": "testTarget",
      "type": "test",
      "versioning": "test",
    },
  },
  {
    "args": {
      "*": [],
      "target": {
        "id": "testTarget",
        "type": "test",
        "versioning": "test",
      },
    },
    "name": "TestProject.getEnvVersioningByTarget",
  },
  {
    "name": "TestProject.getEnvVersioningByTarget",
    "result": {
      "$$data": "TestVersioningService",
      "$$type": "instanceOf",
    },
  },
  {
    "args": {
      "*": [
        "branch",
      ],
      "target": {
        "id": "testTarget",
        "type": "test",
        "versioning": "test",
      },
    },
    "name": "TestVersioningService.getCurrent",
  },
  {
    "name": "TestVersioningService.getCurrent",
    "result": "testBranch",
  },
  {
    "args": {
      "*": [],
      "branch": "testBranch",
      "repo": "test",
    },
    "name": "TestGithubIntegrationService.branchDelete",
  },
  {
    "name": "TestGithubIntegrationService.branchDelete",
  },
]
`;

exports[`Github stream should get state 1`] = `
[
  {
    "args": {
      "*": [],
      "stream": {
        "config": {
          "branch": "branch",
          "integration": "test",
          "org": "org",
        },
        "id": "test",
        "ref": {
          "projectId": "test",
        },
        "type": "github",
      },
    },
    "name": "GithubStreamService.streamGetState",
  },
  {
    "args": {
      "*": [],
      "assertType": "github",
      "stream": {
        "config": {
          "branch": "branch",
          "integration": "test",
          "org": "org",
        },
        "id": "test",
        "ref": {
          "projectId": "test",
        },
        "type": "github",
      },
    },
    "name": "TestProject.getEnvIntegraionByTargetStream",
  },
  {
    "name": "TestProject.getEnvIntegraionByTargetStream",
    "result": {
      "$$data": "TestGithubIntegrationService",
      "$$type": "instanceOf",
    },
  },
  {
    "args": {
      "*": [],
      "assertType": "github",
      "stream": {
        "config": {
          "branch": "branch",
          "integration": "test",
          "org": "org",
        },
        "id": "test",
        "ref": {
          "projectId": "test",
        },
        "type": "github",
      },
    },
    "name": "TestProject.getEnvIntegraionByTargetStream",
  },
  {
    "name": "TestProject.getEnvIntegraionByTargetStream",
    "result": {
      "$$data": "TestGithubIntegrationService",
      "$$type": "instanceOf",
    },
  },
  {
    "args": {
      "*": [],
      "stream": {
        "config": {
          "branch": "branch",
          "integration": "test",
          "org": "org",
        },
        "id": "test",
        "ref": {
          "projectId": "test",
        },
        "type": "github",
      },
    },
    "name": "TestProject.getTargetByTargetStream",
  },
  {
    "name": "TestProject.getTargetByTargetStream",
    "result": {
      "id": "testTarget",
      "type": "test",
      "versioning": "test",
    },
  },
  {
    "args": {
      "*": [],
      "target": {
        "id": "testTarget",
        "type": "test",
        "versioning": "test",
      },
    },
    "name": "TestProject.getEnvVersioningByTarget",
  },
  {
    "name": "TestProject.getEnvVersioningByTarget",
    "result": {
      "$$data": "TestVersioningService",
      "$$type": "instanceOf",
    },
  },
  {
    "args": {
      "*": [
        "branch",
      ],
      "target": {
        "id": "testTarget",
        "type": "test",
        "versioning": "test",
      },
    },
    "name": "TestVersioningService.getCurrent",
  },
  {
    "name": "TestVersioningService.getCurrent",
    "result": "testBranch",
  },
  {
    "args": {
      "*": [],
      "branch": "testBranch",
      "repo": "test",
    },
    "name": "TestGithubIntegrationService.branchGet",
  },
  {
    "name": "TestGithubIntegrationService.branchGet",
    "result": {
      "_links": {
        "html": "branchLink",
      },
      "commit": {
        "commit": {
          "author": {
            "html_url": "commitAuthorLink",
            "name": "commitAuthorName",
          },
          "committer": {
            "date": "2020-01-01",
          },
          "message": "commitMessage",
        },
        "html_url": "commitLink",
        "sha": "commitSha",
      },
    },
  },
  {
    "args": {
      "*": [],
      "stream": {
        "config": {
          "branch": "branch",
          "integration": "test",
          "org": "org",
        },
        "id": "test",
        "ref": {
          "projectId": "test",
        },
        "type": "github",
      },
    },
    "name": "TestProject.getTargetByTargetStream",
  },
  {
    "name": "TestProject.getTargetByTargetStream",
    "result": {
      "id": "testTarget",
      "type": "test",
      "versioning": "test",
    },
  },
  {
    "args": {
      "*": [],
      "target": {
        "id": "testTarget",
        "type": "test",
        "versioning": "test",
      },
    },
    "name": "TestProject.getEnvVersioningByTarget",
  },
  {
    "name": "TestProject.getEnvVersioningByTarget",
    "result": {
      "$$data": "TestVersioningService",
      "$$type": "instanceOf",
    },
  },
  {
    "args": {
      "*": [],
      "branch": "testBranch",
      "org": "org",
      "repo": "test",
    },
    "name": "TestGithubIntegrationService.workflowRunsGet",
  },
  {
    "name": "TestGithubIntegrationService.workflowRunsGet",
    "result": [
      {
        "actor": {
          "html_url": "workflowActorLink",
          "name": "workflowActorName",
        },
        "id": "workflowId",
      },
    ],
  },
  {
    "args": {
      "*": [],
      "org": "org",
      "repo": "test",
      "runId": "workflowId",
    },
    "name": "TestGithubIntegrationService.workflowJobsGet",
  },
  {
    "name": "TestGithubIntegrationService.workflowJobsGet",
    "result": [
      {
        "html_url": "workflowJobLink",
        "steps": [
          {
            "completed_at": "2020-01-02",
            "conclusion": "success",
            "name": "workflowJobName",
            "number": 1,
            "started_at": "2020-01-01",
          },
        ],
      },
    ],
  },
  {
    "name": "GithubStreamService.streamGetState",
    "result": {
      "history": {
        "action": [
          {
            "author": {
              "link": "workflowActorLink",
              "name": "workflowActorName",
            },
            "description": undefined,
            "id": "workflowId",
            "link": null,
            "metadata": {},
            "status": null,
            "steps": {
              "1": {
                "description": "workflowJobName",
                "id": "1",
                "link": "workflowJobLink#step:1:1",
                "runningTimeSeconds": 86400,
                "status": "completed",
                "type": "github:workflow:job",
              },
            },
            "time": null,
            "type": "github:workflow",
          },
        ],
        "artifact": [],
        "change": [
          {
            "author": {
              "link": "commitAuthorLink",
              "name": "commitAuthorName",
            },
            "description": "commitMessage",
            "id": "commitSha",
            "link": "commitLink",
            "metadata": {},
            "status": null,
            "steps": null,
            "time": {},
            "type": "github:commit",
          },
        ],
      },
      "id": "test",
      "isSyncing": false,
      "link": "branchLink",
      "metadata": {
        "branch": "testBranch",
        "org": "org",
      },
      "projectId": "test",
      "targetId": undefined,
      "type": "github",
      "ver": 1,
      "version": null,
    },
  },
]
`;
